rfcs/text/0013_background_sessions.md

```diff
@@ -0,0 +1,482 @@
+}
+```
+
+## Frontend Services
+
+Most sessions will probably not be saved. Therefore, to avoid creating unnecessary saved objects, the browser will keep track of requests and their respective search IDs, until the user chooses to store the session. Once a session is stored, any additional searches will be immediately saved on the server side. 
+
+### New Session Service
+
+We will expose a new frontend `session` service on the `data` plugin `search` service. 
+
+The service will expose the following APIs:
+
+```ts
+interface ISessionService {
+   /** 
+    * Returns the current session ID
+    */
+   getId: () => string;
+
+   /**
+    * Start a new session, by generating a new session ID.
+    */
+   start: () => string;
+
+   /**
+    * Restore a session. Any searches executed using this ID will attempt to reload saved data.
+    * A restored session may be expired, meaning that following search requests will fail. 
+    * @param sessionId Session ID to restore.`.
+    */
+   restore: (sessionId: string) => void;
+
+   /**
+    * Close the current session and start a new one
+    */
+   clear: () => void; 
+
+   /**
+    * Store a session.
+    * @param sessionId Session ID to store. Probably retrieved from `sessionService.get()`.
+    * @param name A display name for the session.
+    * @param url TODO: is the URL provided here? How?
+    * @param searchIdMapping An optional mapping of hashed requests mapped to the corresponding searchId.
+    * @returns The stored `BackgroundSessionAttributes` object
+    * @throws Throws an error in OSS.
+    */
+   store: (sessionId: string, name: string, url: string, searchIdMapping?: Record<string, string>)) => Promise<BackgroundSessionAttributes>
+
+   /**
+    * @returns Is the current session stored (i.e. is there a saved object correspondng with this sessionId).
+    */
+   isStored: () => boolean;
+
+   /** 
+    * @returns Is the current session restored (i.e. was loaded by calling `sessionService.restore()`).
+    */
+   isRestore: () => boolean;
+
+   /**
+    * Mark a session as and all associated searchIds as expired.
+    * Cancels active requests, if there are any.
+    * @param sessionId Session ID to store. Probably retrieved from `sessionService.get()`.
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+   expire: (sessionId: string) => Promise<boolean> 
+   
+   /**
+    * Extend a session and all associated searchIds.
+    * @param sessionId Session ID to extend. Probably retrieved from `sessionService.get()`.
+    * @param extendBy Time to extend by, can be a relative or absolute string.
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+   extend: (sessionId: string, extendBy: string)=> Promise<boolean>
+
+   /**
+    * @param TODO: input arguments
+    * @returns a filtered list of BackgroundSessionAttributes objects.
+    * @throws Throws an error in OSS.
+    */
+   list: () => BackgroundSessionAttributes[]
```

Comment:
Add same parameters as whatever the saved object client accepts.

Reply:
Since background sessions are scoped to the current user (mostly?), they will need to be registered as a `hidden` saved object type, so that they aren't accessible via the standard `/api/saved_objects/*` endpoints.

Reply:
@legrego Yes, that's the plan. I can update the RFC to explain that.

Reply:
@legrego If we are using the `hidden` object type, does this mean we won't be able to use the standard saved objects client to retrieve a list of background sessions? Or is there a `hidden` parameter we can send in the search request?

Reply:
> If we are using the hidden object type, does this mean we won't be able to use the standard saved objects client to retrieve a list of background sessions? Or is there a hidden parameter we can send in the search request?

You will need to construct a saved objects client that's specifically configured to show this hidden type via the `includedHiddenTypes` configuration option. Here's an example from an API test:

https://github.com/elastic/kibana/blob/bf04235dae35452061cc7ea3d86d96c19a58206c/x-pack/test/encrypted_saved_objects_api_integration/fixtures/api_consumer_plugin/server/hidden_saved_object_routes.ts#L101-L105.